### PR TITLE
Initialization of TypeIdentifier is not thread-safe

### PR DESCRIPTION
**Pull request by jrw972**
Problem
-------

Initialization of statics is not thread-safe on all compilers.  See #2362.

Solution
--------

Add lock to serialize initialization of type identifier.

### CODE REVIEW

**Review by ClaytonCalabrese — COMMENTED** [review]
Do we want a different lock for this or is the TypeMap one fine?

**Comment by jrw972** [comment]
I'll defer to @mitza-oci  on that one.

**Comment by iguessthislldo** [comment]
> Initialization of statics is not thread-safe on all compilers. See

See what?

File: dds/idl/typeobject_generator.cpp [hunk 1]
```diff
@@ -1762,7 +1762,11 @@ typeobject_generator::generate(AST_Type* node, UTL_ScopedName* name)
     gti.endArgs();
     const OpenDDS::XTypes::TypeIdentifier ti = get_minimal_type_identifier(node);
     be_global->impl_ <<
-      "  static const XTypes::TypeIdentifier ti = " << ti << ";\n"
+      "  static XTypes::TypeIdentifier ti;\n"
+      "  ACE_GUARD_RETURN(ACE_Thread_Mutex, guard, TheServiceParticipant->get_tm_lock(), ti);\n"
+      "  if (ti.kind() == XTypes::TK_NONE) {\n"
```
Comment (mitza-oci) [inline dds/idl/typeobject_generator.cpp:1767]: Move lock within the `if`?

Comment (jrw972) [reply dds/idl/typeobject_generator.cpp:1767]: One thread could return a partially constructed type identifier if the lock is within the `if`.  If anything, the lock should be first.  I'll fix the other instance as well.

Comment (mitza-oci) [reply dds/idl/typeobject_generator.cpp]: Sorry, I misread it.

File: dds/idl/typeobject_generator.cpp [hunk 2]
```diff
@@ -1762,7 +1762,11 @@ typeobject_generator::generate(AST_Type* node, UTL_ScopedName* name)
     gti.endArgs();
     const OpenDDS::XTypes::TypeIdentifier ti = get_minimal_type_identifier(node);
     be_global->impl_ <<
-      "  static const XTypes::TypeIdentifier ti = " << ti << ";\n"
+      "  ACE_GUARD_RETURN(ACE_Thread_Mutex, guard, TheServiceParticipant->get_tm_lock(), ti);\n"
+      "  static XTypes::TypeIdentifier ti;\n"
```
Comment (mitza-oci) [inline dds/idl/typeobject_generator.cpp:1766]: Now this looks like a problem because (in the generated code) we have ti referenced before it's defined.

**Comment by mitza-oci** [comment]
> 
> 
> > Initialization of statics is not thread-safe on all compilers. See
> 
> See what?

Looks like it was updated after the initial post.

**Comment by mitza-oci** [comment]
> 
> 
> I'll defer to @mitza-oci on that one.

I think it's fine to have a single lock handling this initialization.  We don't expect it to be in the critical path of data transmission, it should just be at startup.  So the goal here is for correctness on all platforms, not necessarily maximum scalability.

The name `get_tm_lock` and comment around it in Service_Participant should be updated.

**Review by mitza-oci — APPROVED** [review]

**Review by ClaytonCalabrese — APPROVED** [review]

**Pull request merged**